crates/provider/src/wallet.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{
+    fillers::{FillProvider, JoinFill, SignerFiller, TxFiller},
+    Provider,
+};
+use alloy_network::{Ethereum, Network, NetworkSigner};
+use alloy_primitives::Address;
+use alloy_transport::Transport;
+
+/// Trait for Providers, Fill stacks, etc, which contain [`NetworkSigner`].
+pub trait WalletProvider<N: Network = Ethereum> {
+    /// The underlying [`NetworkSigner`] type contained in this stack.
+    type Signer: NetworkSigner<N>;
+
+    /// Get a reference to the underlying signer.
+    fn signer(&self) -> &Self::Signer;
+
+    /// Get a mutable reference to the underlying signer.
+    fn signer_mut(&mut self) -> &mut Self::Signer;
+
+    /// Get the default signer address.
+    fn default_signer(&self) -> Address {
+        self.signer().default_signer()
+    }
+
+    /// Check if the signer can sign for the given address.
+    fn is_signer_for(&self, address: &Address) -> bool {
+        self.signer().is_signer_for(address)
+    }
+
+    /// Get an iterator of all signer addresses.
```

Comment:
should mention that this is non-empty because always includes the default signer address